### ISSUE TITLE
Deadlock in rate limiter

### ISSUE THREAD

**Issue by benmoss** [buildkite/agent-stack-k8s#106]
#90 introduced a different way of managing max-in-flight. We saw that the old way was buggy and would allow more jobs in flight than the user specified, because there's lag in between the job being created and the cached informer learning about it.

The new rate limiter has a couple bugs I noticed, but the worst is a deadlock. [This](https://gist.github.com/benmoss/9c6f9dd6850145319b82675a3af49b45) is the stack dump, basically it's stuck [waiting on a `completion`](https://github.com/buildkite/agent-stack-k8s/blob/f9cfa80c3e70452e717ffa411f7be2f3fc1e0fce/scheduler/limiter.go#L134), holding the mutex, while at the same time it's trying to [add a new job](https://github.com/buildkite/agent-stack-k8s/blob/f9cfa80c3e70452e717ffa411f7be2f3fc1e0fce/scheduler/limiter.go#L90).

Not sure I can totally reason about how completions is at capacity, but at least one reason is that we're not handling [OnAdd](https://github.com/buildkite/agent-stack-k8s/blob/f9cfa80c3e70452e717ffa411f7be2f3fc1e0fce/scheduler/limiter.go#L105), which means that when the controller starts it ignores any jobs that are already running. We then get completions for those jobs when they finish/delete, meaning we might have more completions than we have jobs in flight.

Another problem is the check for `job.Status.CompletionTime`, you only get that if a job "completes", aka finishes successfully.
> The completion time is only set when the job finishes successfully
It seems there's no single way to find if a job just finished, we need to look at multiple fields, or multiple conditions, to get both success or failure.

Another problem is that if the limiter sends a job on to the scheduler but the scheduler errors when trying to create it, it will never get removed from the in-flight map, since we have no way of learning about creating failures.

---
Basically, a bunch of problems. I think the right solution is to
- Change the API to just be a regular middleware pattern, like [http.Handler](https://pkg.go.dev/net/http#Handler). Gets rid of the nonsense of channels on the external API, and allows errors to bubble back up through the Limiter
- Just do a `len(inFlight)` to check if we're at max-in-flight, and if so we can just drop the Job and let it be requeued later when we poll the API again. Get rid of the `completions` channel and its over complicated attempt to do this rate-limiting.
- Fix up the job completion logic, that part is pretty straightforward
